cmd/ctr/commands/images/images.go

```diff
@@ -355,3 +357,77 @@ var removeCommand = cli.Command{
 		return exitErr
 	},
 }
+
+var pruneCommand = cli.Command{
+	Name:  "prune",
+	Usage: "remove unused images",
+	Flags: []cli.Flag{
+		cli.BoolFlag{
+			Name:  "all", // TODO: add more filters
+			Usage: "remove all unused images, not just dangling ones (if all is not specified no images will be pruned)",
+		},
+	},
+	// adapted from `nerdctl`:
+	// https://github.com/containerd/nerdctl/blob/272dc9c29fc1434839d3ec63194d7efa24d7c0ef/cmd/nerdctl/image_prune.go#L86
+	Action: func(context *cli.Context) error {
+		client, ctx, cancel, err := commands.NewClient(context)
+		if err != nil {
+			return err
+		}
+		defer cancel()
+
+		all := context.Bool("all")
+		if !all {
+			log.G(ctx).Warn("No images pruned. `image prune` requires --all to be specified.")
+			// NOP
+			return nil
+		}
+
+		var (
+			imageStore     = client.ImageService()
+			contentStore   = client.ContentStore()
+			containerStore = client.ContainerService()
+		)
+		imageList, err := imageStore.List(ctx)
+		if err != nil {
+			return err
+		}
+		containerList, err := containerStore.List(ctx)
+		if err != nil {
+			return err
+		}
+		usedImages := make(map[string]struct{})
+		for _, container := range containerList {
+			usedImages[container.Image] = struct{}{}
+		}
+
+		delOpts := []images.DeleteOpt{images.SynchronousDelete()}
+		removedImages := make(map[string][]digest.Digest)
+		for _, image := range imageList {
+			if _, ok := usedImages[image.Name]; ok {
+				continue
+			}
+
+			digests, err := image.RootFS(ctx, contentStore, platforms.DefaultStrict())
+			if err != nil {
+				log.G(ctx).WithError(err).Warnf("failed to enumerate rootfs of image %s", image.Name)
+			}
+			if err := imageStore.Delete(ctx, image.Name, delOpts...); err != nil {
```

Comment:
If the delete could be deferred to the next loop, `SynchronousDelete` option could be reserved for the last run. Synchronous delete can also be run for a lease as well. It is not a good pattern to do synchronous deletes in a loop though.

Reply:
Thanks for the suggestion. I updated to use a `delCount=imageCount - usedImageCount` and add `SynchronousDelete` when deleting the last one `delCount=0`. 

It does have an assumption that all used images by containers (`container.Image`) appear in `imageList`, so that the `delCount=imageCount - usedImageCount` is correct.

If the assumption doesn't hold, I can change to first `for loop imageList` to count (or store in a slice) the `image` to be deleted. Then delete them in another for-loop.

https://github.com/containerd/containerd/pull/7730/files#diff-5c91be0697d8236c280ce6a166ffe908feeffec3d246d5dad47ba4bf665e89a9R411-R414

Reply:
not sure if it helps .. but here is the code in CRI for removing an image:

https://github.com/containerd/containerd/blame/main/pkg/cri/server/image_remove.go#L36-L68

Reply:
Hi Derek, thanks for the explanation! I've updated the logic to using `SynchronousDelete` only for the last run (and ignore `IsNotFound` err).